### PR TITLE
[ENH] in `torch` based TSC and TSR, add support for evaluation metrics from `torchmetrics`

### LINKED ISSUE #9329: [ENH] Add support for different evaluation metrics in torch deep learning models
## Is your feature request related to a problem? Please describe.
Currently in all the PyTorch-based estimators only accuracy is supported as an evaluation metric.
Whereas in TensorFlow-based estimators, different metrics can be passed  in `model.compile` as an argument.

Add support for multiple metrics in PyTorch-based estimators.

## Describe the solution you'd like.
While PyTorch doesn't have something in-built akin to TensorFlow, there other libraries from the ecosystem like `torchmetrics` which can be utilised here.
Few considerations are mentioned below, a recommendation needs to be made and then corresponding choice implemented.

## Additional context
- `torchmetrics` is the defacto choice these days.
- The official `torcheval` is a lightweight package.
- Darts has a dedicated `metrics` module for time series.
- Any other popular choices?
- Technically`sklearn.metrics` can also be used with PyTorch models but the PyTorch outputs must first be converted to NumPy arrays on the CPU. Direct use of PyTorch tensors with scikit-learn is not natively supported. So it's not a great choice.

### PR DESCRIPTION
**Pull request by Faakhir30**
#### Reference Issues/PRs
Fixes #9329 

#### What does this implement/fix? Explain your changes.
- adds torchmetrics support to `BaseDeepRegressorTorch` and `BaseDeepClassifierTorch`
- Add `metrics` param support to existing classifiers and regressors.

#### Does your contribution introduce a new dependency? If yes, which one?

Yes, `torchmetrics` is used as a soft dependency.
 
#### What should a reviewer concentrate their feedback on?

- Are all  common metrics provided by `torchmetrics`  object useable under current implementation. (I'm checking this also right now)
- I have not added it to `python_dependencies` tag, as it will check for all instantiations of regressors, while this soft dep is not required in this module always, so raising error in `__init__` myself, instead of adding it to tags.

#### Did you add any tests for the change?

No, not yet.

#### Any other comments?

- I see there are no tests for `BaseDeepRegressorTorch` specifically at `test_base.py` level, should we have test for this metrics thing?

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Would it not be better to expect the user to pass the metrics instances directly, rather than string reference and parameters?

**Comment by Faakhir30** [comment]
> Would it not be better to expect the user to pass the metrics instances directly, rather than string reference and parameters?

Yes, I am allowing to pass `Callable` (docstring needed an update, done). 

`str` might be easy to use when user wants to use some common ones, like `meansquarederror` or `r2`. 

If you would like `str` completly removed, lmk @fkiraly .
This `metrics` needs to be carried forward to params and docstrings of classifiers and regressors.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

* I do not see the reasoning behind constructing with string alasess. This has the complication that you have to pass kwargs separately, it causes that you cannot pass arbitrary metrics (you have to hard-code them), and it causes another instance of the "manual string resolution pattern" that you have in other pull requests. I would simply pass the `torchmetrics` classes.
    * if you insist on strings, I would do as follows: class names map to default parameter settings; for parameter definition, you can pass a `craft`-like string. But I think this might overcomplicate the present PR, perhaps something for a separate PR.
* to ensure this can be tested, you can use the `tests:python_dependencies` tag to ensure `torchmetrics` is also present in tests for some estimators.

**Comment by Faakhir30** [comment]
> if you insist on strings, I would do as follows: class names map to default parameter settings; for parameter definition, you can pass a craft-like string. But I think this might overcomplicate the present PR, perhaps something for a separate PR.

I'd like to go with this middle ground (without `craft` for now) of having strings with auto/default parameter settings, and users can pass a callable if a metric with custom kwargs is needed. This approach:

* removes `metrics_kwargs` from the constructor and brings clarity to using a callable if custom kwargs are needed.
* keeps the API design open for a `craft`-based approach if needed down the line. Though I'd also prefer to keep it simple with the above specification: string (default kwargs used) + callable (for custom kwargs).

cc @fkiraly

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

Please do not put dependency checks in `__init__` - you can move this check to `__post_init__`.

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 1]
```diff
@@ -156,13 +180,23 @@ def _fit(self, X, y):
...
             y_pred = self.network(**inputs)
             loss = self._criterion(y_pred, outputs)
             self._optimizer.zero_grad()
             loss.backward()
             self._optimizer.step()
             losses.append(loss.item())
+
+            # Compute metrics if any
+            if self._metrics_objects:
+                with _safe_import("torch.no_grad")():
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base/_base_torch.py]: no `_safe_import` needed here, because when we reach this place, we can assume `torch` is present.

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 2]
```diff
@@ -527,6 +567,83 @@ def _instantiate_criterion(self):
...
+        ------
+        ValueError
+            If an unknown metric name is passed.
+        TypeError
+            If metric is neither a string nor a callable.
+        """
+        if self.metrics is None:
+            return None
+
+        torchmetrics = _safe_import("torchmetrics")
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base/_base_torch.py:750]: this may cause unexpected behaviour of `torchmetrics` is not present.

Why not gate it behind a check, and then assume `torchmetric` is present?

Comment (Faakhir30) [reply sktime/classification/deep_learning/base/_base_torch.py:750]: Placed a dependency check in `post_init` (previously was in `init`), so it seems unnecessary to check again in `_fit`. Assuming `fit` only gets called after `post_init`.

in `post_init`:

```python
        if self.metrics:
            _check_soft_dependencies("torchmetrics", severity="error")
```

**Comment by fkiraly** [comment]
@Faakhir30, thanks! Minor request, please do not move tickets on the board, only in the stand-up, so everyone is aware.
(I should also stop moving tickets outside the stand-up)

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 3]
```diff
@@ -121,6 +130,9 @@ def __post_init__(self):
             torchManual_seed = _safe_import("torch.manual_seed")
             torchManual_seed(self.random_state)
 
+        if self.metrics:
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base/_base_torch.py]: this should not be here. It should be in `__dynamic_tags__`, and set the tag to include `torchmetrics`, instead of checking.

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 4]
```diff
@@ -136,8 +148,13 @@ def __post_init__(self):
 
         # use this when y has str
         self.label_encoder = None
+        self._metrics_objects = None
 
     def _fit(self, X, y):
+        if self.random_state is not None:
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base/_base_torch.py:203]: you can assume that `torch` is present in `_fit`.

File: sktime/classification/deep_learning/base/_base_torch.py (hunk 2 above)
Comment (fkiraly) [reply sktime/classification/deep_learning/base/_base_torch.py:750]: see above how to handle it

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 5]
```diff
@@ -527,6 +567,63 @@ def _instantiate_criterion(self):
...
+            if isinstance(metric, str):
+                if hasattr(torchmetrics, metric):
+                    metric_class = getattr(torchmetrics, metric)
+                    kwargs = {"task": "multiclass", "num_classes": self.num_classes}
+                    if metric in ("F1Score", "Precision", "Recall"):
+                        kwargs["average"] = "macro"
+                    metrics_dict[metric] = metric_class(**kwargs)
+                else:
+                    raise ValueError(
+                        f"Unknown metric: {metric}. Please pass one of the available "
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base/_base_torch.py]: when raising a `ValueError`, remember this best practice advice: at the start say which class or method raised the error. Reason, users might have complex pipelines, so it is better to tell them where exactly the error is coming from.

E.g., `f"Error in constructing torch based classifier {type(self).__name__}, unknown metric:"`

**Comment by fkiraly** [comment]
A comment rather than a change request to this PR: there seems to be substantial duplication between the torch based classifier and regressor classes. I would suggest a separate PR, now that you know these classes well: move the commonalities to a common base class (put e.g., in `sktime.base`), and inherit both from that.

**Comment by Faakhir30** [comment]
Failing tests are irrelvent to this PR, its related to `PytorchForecastingTFT` docstring examples not having expected "doctest skip" comments.

https://github.com/sktime/sktime/actions/runs/26751617867/job/78841195051?pr=9390#step:10:651

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 6]
```diff
@@ -527,6 +572,65 @@ def _instantiate_criterion(self):
...
+        metrics_dict = {}
+
+        for metric in metrics_list:
+            if isinstance(metric, str):
+                if hasattr(torchmetrics, metric):
+                    metric_class = getattr(torchmetrics, metric)
+                    kwargs = {"task": "multiclass", "num_classes": self.num_classes}
+                    if metric in ("F1Score", "Precision", "Recall"):
+                        kwargs["average"] = "macro"
+                    metrics_dict[metric] = metric_class(**kwargs)
```
Comment (phoeenniixx) [inline sktime/classification/deep_learning/base/_base_torch.py]: I feel this is "too much" indentation  due to nested `if-else`- if we there is more than 4 tabs identation, there are signs that we might need a new method.. to keep this clean.

**Review by phoeenniixx — CHANGES_REQUESTED** [review]
I think we should avoid using too much nested conditions and try to move them to another method or simplify the conditions

File: sktime/classification/deep_learning/base/_base_torch.py (hunk 6 above)
Comment (Faakhir30) [reply sktime/classification/deep_learning/base/_base_torch.py]: unable to create meaningfull seperate function, I usually avoid creating a function if I can't give it a name that clarifies that methods responsibility.

So, simplified conditions, reduced indentation to 3 levels ;)

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 7]
```diff
@@ -527,6 +572,64 @@ def _instantiate_criterion(self):
                 f"But got {type(self._validated_criterion)} instead."
             )
 
+    def _instantiate_metrics(self):
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base/_base_torch.py]: make this a method that takes `metrics` as an argument, as opposed to obtaining from `self`.

General best practice: if a method can easily be "pure", i.e., not read from `self` or have side effects on it, make it pure.

Related, this method is badly documented since you do not document what the assumptions are, e.g., what `self.metrics` can be.

Make sure you document that as a "parameter" once you make the function pure.

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 8]
```diff
@@ -147,6 +169,11 @@ def _fit(self, X, y):
         self._optimizer = self._instantiate_optimizer()
         # instantiate callbacks (learning rate schedulers)
         self._schedulers = self._instantiate_schedulers()
+        # ensure num_classes is set before instantiating metrics
+        # as classification metrics require num_classes as an argument
+        self.num_classes = len(np.unique(y))
+        # instantiate metrics
+        self._metrics_objects = self._instantiate_metrics()
```
Comment (fkiraly) [inline sktime/classification/deep_learning/base/_base_torch.py]: and I would move this to `__post_init__`.

**Review by fkiraly — CHANGES_REQUESTED** [review]
See above, I would make the logic clearer, and couple the call of the metrics coercion to only one location.

File: sktime/classification/deep_learning/base/_base_torch.py (hunk 8 above)
Comment (Faakhir30) [reply sktime/classification/deep_learning/base/_base_torch.py]: - metric objects are statefull, that means for all metrics we need to call `metric.reset()` before each `fit` (if we move them to **post_init**), we can do reset same object approach, but another issue is:
- classification metrics require `num_classes` as argument, and `self.num_classes` is only set after `build_network` calll in `fit`, so its not possible to move metrics initialization to `post_init`.

File: sktime/classification/deep_learning/base/_base_torch.py [hunk 9]
```diff
@@ -527,6 +574,76 @@ def _instantiate_criterion(self):
...
+                metrics_dict[metric] = metric_class(**kwargs)
+            elif isinstance(metric, Callable):
+                metric_name = metric.__class__.__name__
+                metrics_dict[metric_name] = metric
+            else:
+                raise TypeError(
+                    "`metrics` can either be None, a str or a tuple of str "
+                    "representing metrics from torchmetrics, or an instance of a "
+                    f"torchmetrics metric. But got {type(metric)} instead."
+                )
```
Comment (phoeenniixx) [inline sktime/classification/deep_learning/base/_base_torch.py]: I think this method can be moved to a new method (like `_get_metrics`) or similar.
Having so much indentation means we should create a new method

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

The test in `test_all_estimators` should not be modified.
Instead, a plugin to `deep_equals` should be added to handle `torchmetrics`.

For this, modify the `plugins_default` in `sktime.utils.deep_equals`.

File: sktime/classification/deep_learning/base/_base_torch.py (hunk 9 above)
Comment (fkiraly) [reply sktime/classification/deep_learning/base/_base_torch.py]: yes, exactly. Same principle as in `_instantiate_activations`.

File: sktime/classification/deep_learning/base/_base_torch.py (hunk 8 above)
Comment (fkiraly) [reply sktime/classification/deep_learning/base/_base_torch.py]: ok, I see. I would say, make `_instantiate_metrics` a pure function then.

File: sktime/utils/deep_equals/_deep_equals.py [hunk 10]
```diff
@@ -279,3 +283,70 @@ def _polars_equals_plugin(x, y, return_msg=False):
...
+        hasattr(x, "_defaults")
+        and hasattr(x, "reset")
+        and hasattr(type(x), "__module__")
+        and type(x).__module__.startswith("torchmetrics")
+    ):
+        return None
+
+    from sktime.utils.dependencies import _check_soft_dependencies, _safe_import
+
+    if not _check_soft_dependencies("torchmetrics", severity="none"):
```
Comment (fkiraly) [inline sktime/utils/deep_equals/_deep_equals.py]: just as a safety consideration, I think it is safer to do the dependency check first. Since - theoretically - `module` might not have `startswith` as an attribute/method.

**Review by fkiraly — DISMISSED** [review]
Looks good to me now, remaining changes are non-blocking.

File: sktime/classification/deep_learning/base/_base_torch.py (hunk 9 above)
Comment (Faakhir30) [reply sktime/classification/deep_learning/base/_base_torch.py]: created another method for indivisual metric instantiation, done, thanks

**Review by fkiraly — APPROVED** [review]
